src/interface/update.cpp

```diff
@@ -46,7 +47,9 @@ TaskStatus FluxDivergence(MeshBlockData<Real> *in, MeshBlockData<Real> *dudt_con
   pmb->par_for(
       "FluxDivergenceBlock", 0, vin.GetDim(4) - 1, kb.s, kb.e, jb.s, jb.e, ib.s, ib.e,
       KOKKOS_LAMBDA(const int l, const int k, const int j, const int i) {
-        dudt(l, k, j, i) = FluxDivHelper(l, k, j, i, ndim, coords, vin);
+        if (dudt.IsAllocated(l) && vin.IsAllocated(l)) {
```

Comment:
This is a very good question. My first thought is that I'm not really worried about thread divergence here, because the threads where the condition is false, will simply not do anything, but they won't be doing anything else that needs to be serialized. However, it is true that we're wasting some resources, because there will be a bunch of threads that just check the condition and do nothing else. So it would be nice to optimize this.

I was thinking of adding a feature to the packing functions so that one can exclude unallocated variables, however, this could lead to variable index mismatches between packs and it's not clear how we could even check for that.

So I like @pgrete's suggestion the best. We can add a function that basically computes the intersection of the allocated indices between multiple packs and then we only loop over those indices. Integrating that into `par_for` would certainly be beyond this PR, but I'll see if I can add a simple function to get that union of indices and loop over those.

Reply:
@jdolence @pgrete This took way longer than I thought. While implementing a possible solution, I found a very nasty bug that resulted in illegal memory accesses on CUDA and it took way too long to track down. But it's fixed now. Here is a possible implementation: https://github.com/lanl/parthenon/commit/8472b5d2160c9583925fd1aa184b7d708c52b5b0

I like the interface, but unfortunately the performance sucks :( Compared to just calling `IsAllocated` in the inner loops, the above commit with using `AllocatedIndices` is almost 2x slower. Almost certainly because of the host-to-device copies of the indices. So I think we should either just keep the `IsAllocated` checks, or maybe switch the loops to hierarchical parallelism and just call `IsAllocated` in the outer most loop. That approach would not result in additional host-to-device transfers, but I have a feeling it won't perform much better than calling `IsAllocated` in the inner loop since that's very cheap already (but we won't know for sure until we try).